File: scripts/paper/paper_create.js
const OpenAI = require('openai');

const fs = require('fs');
const crypto = require('crypto');

const client = new OpenAI({
  apiKey: '',
  //sk-7nVZaP1jPhutqnHnDhGmT
  //3BlbkFJoFoK3iAoOiUZp0ES7r87
});


function generatePrompts(fileContent) {
  const data = JSON.parse(fileContent);
  const title = data.title;
  const messages = {};
  const prompts = [];
  const role = 'user'
  let index = 0;
  data.outline.forEach((sectionData, sectionIndex) => {
    const section = sectionData.section;
    sectionData.content.forEach((contentTitle, contentIndex) => {
      const prompt = `你现在是一个论文写作助理，我需要写一篇毕业论文，题目为：《${title}》，请帮我完成 章节 “${section}:${contentTitle}” 的 内容，字数300字，并附上引用文档链接。`;
      const timestamp = Date.now();
      const hash = crypto.createHash('md5').update(`user${timestamp}${sectionIndex}${contentIndex}`).digest('hex');
      const id = `${hash}${timestamp}${sectionIndex}${contentIndex}`;
      prompts.push({ content: prompt, role });
      messages[index]={ id, content: prompt, role,section,contentTitle,title };
      index++
    });
  });
  return { prompts, messages };
}

const filePath = './titles.json';
const fileContent = fs.readFileSync(filePath, 'utf8');
const { prompts, messages } = generatePrompts(fileContent);

let word_len = 0
async function processMessageOne() {
  let index = 0;

  while (prompts.length > 0) {
    const message_one = [prompts.shift()];
    const stream = await client.beta.chat.completions.stream({
      model: 'gpt-3.5-turbo-0613', //gpt-3.5-turbo
      messages: message_one,
      stream: true,
    });

    let message_info = messages[index]
    let {section,contentTitle,title} = message_info
    console.log(`\n论文《${title}》-> ${section} -> ${contentTitle}: \n`)
    stream.on('content', (delta, snapshot) => {
      process.stdout.write(delta);
    });

    let finalContent = '';
    for await (const chunk of stream) {
      finalContent += chunk.choices[0]?.delta?.content || '';
      word_len++
    }

    const chatCompletion = await stream.finalChatCompletion();
    console.log(`chatCompletion`);
    console.log(chatCompletion);
    console.log(`已生成字数: ${word_len}`)
    let con = `\n
${section}:\n
\t${contentTitle}:\n
\t\t${finalContent}\n
    `
    fs.appendFileSync(`chat_output.txt`, con);
    fs.writeFileSync(`chat_output_${index}.txt`, finalContent);
    index++;
  }
}
processMessageOne();



// async function main() {
//   const runner = client.beta.chat.completions
//     .runFunctions({
//       model: 'gpt-3.5-turbo',
//       messages:prompts,
//       functions: [
//         {
//           function: getCurrentLocation,
//           parameters: { type: 'object', properties: {} },
//         },
//         {
//           function: getWeather,
//           parse: JSON.parse,
//           parameters: {
//             type: 'object',
//             properties: {
//               location: { type: 'string' },
//             },
//           },
//         },
//       ],
//     })
//     .on('message', (message) => console.log(message));

//   const finalContent = await runner.finalContent();
//   console.log();
//   console.log('Final content:', finalContent);
// }

// async function getCurrentLocation() {
//   return 'Boston'; // 模拟查找
// }

// async function getWeather(args) {
//   const { location } = args;
//   return { location };
// }

// main();

File: scripts/paper/paper_create.py
import json
from pprint import pprint
import hashlib
import time
import asyncio
# import openai
from openai import AsyncOpenAI
from pathlib import Path

def generate_prompts(file_content):
<<<<<<< HEAD
    # 解析 JSON 内容
    data = json.loads(file_content)

    # 提取标题
    title = data['title']

    # 迭代处理大纲部分
=======
    data = json.loads(file_content)

    title = data['title']

>>>>>>> origin/main
    prompts = []
    for section_index, section_data in enumerate(data['outline']):
        section = section_data['section']
        for content_index, content_title in enumerate(section_data['content']):
<<<<<<< HEAD
            # 生成提示
            prompt = f"根据之前的《{title}》论文：第 {section} 章 -> {content_title}，请生成 {content_title} 内容，字数300字，并附上配图及引用文档链接。"
            # 生成唯一 ID
=======
            prompt = f"根据之前的《{title}》论文：第 {section} 章 -> {content_title}，请生成 {content_title} 内容，字数300字，并附上配图及引用文档链接。"
>>>>>>> origin/main
            timestamp = int(time.time())
            hash_id = hashlib.md5(f"user{timestamp}{section_index}{content_index}".encode()).hexdigest()
            id = f"{hash_id}{timestamp}{section_index}{content_index}"
            prompts.append({"id": id, "content": prompt})
    return prompts

<<<<<<< HEAD
# 读取 JSON 文件
=======
>>>>>>> origin/main
file_path = './a.json'
with open(file_path, 'r', encoding='utf-8') as file:
    file_content = file.read()

messages_queue = generate_prompts(file_content)
pprint(messages_queue)
#
#
#
#
# # 初始化 OpenAI 客户端
<<<<<<< HEAD
client = AsyncOpenAI(api_key="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")
=======
client = AsyncOpenAI(api_key="")#
# sk-7nVZaP1jPhutqnHnDhGm
# T3BlbkFJoFoK3iAoOiUZp0ES7r87
>>>>>>> origin/main

#
# async def handle_message(message):
#     """处理单个消息，并保存响应"""
#     response = await client.chat.completions.create(
#         messages=[
#             {"role": "user", "content": message["content"]}
#         ],
#         model="gpt-3.5-turbo",
#     )
#     # 将响应保存为文本文件
#     Path(f"{message['id']}.txt").write_text(response.choices[0].message.content)
#
# async def main():
#     """主函数，处理消息队列"""
#     for message in messages_queue:
#         await handle_message(message)
#
# # 运行主函数
# asyncio.run(main())
#
#
# pprint(messages_queue)
#